daemon/graphdriver/overlayutils/overlayutils.go

```diff
@@ -55,28 +64,76 @@ func SupportsOverlay(d string, checkMultipleLowers bool) error {
 	}
 	defer func() {
 		if err := os.RemoveAll(td); err != nil {
-			log.G(context.TODO()).Warnf("Failed to remove check directory %v: %v", td, err)
+			log.G(context.TODO()).WithError(err).Warnf("failed to remove check directory %v", td)
 		}
 	}()
 
-	for _, dir := range []string{"lower1", "lower2", "upper", "work", "merged"} {
-		if err := os.Mkdir(filepath.Join(td, dir), 0o755); err != nil {
+	l1, l2, l3, work, merged := filepath.Join(td, "l1"), filepath.Join(td, "l2"), filepath.Join(td, "l3"), filepath.Join(td, "work"), filepath.Join(td, "merged")
+	for _, dir := range []string{l1, l2, l3, work, merged} {
+		if err := os.Mkdir(dir, 0o755); err != nil {
+			return err
+		}
+	}
+
+	// Create files, each holding their own path, which will be used for subsequent testing.
+	f1, f2 := filepath.Join(l1, "f1"), filepath.Join(l1, "f2")
+	for _, file := range []string{f1, f2} {
+		if err := os.WriteFile(file, []byte(file), 0o644); err != nil {
 			return err
 		}
 	}
 
-	mnt := filepath.Join(td, "merged")
-	lowerDir := path.Join(td, "lower2")
+	// RENAME_EXCHANGE will cause an atomic swap of f1 and f2. Create two files, each with their own paths as their
+	// contents.
+	err = unix.Renameat2(0, f1, 0, f2, unix.RENAME_EXCHANGE)
+	if err != nil {
+		return err
+	}
+	// The contents of f1 should now be the path to f2, and vice versa.
+	content, err := os.ReadFile(f1)
+	if !bytes.Equal(content, []byte(f2)) || err != nil {
+		return fmt.Errorf("RENAME_EXCHANGE of %v and %v failed: %w", f1, f2, err)
+	}
+
+	// RENAME_NOREPLACE should cause an error since f2 already exists.
+	err = unix.Renameat2(0, f1, 0, f2, unix.RENAME_NOREPLACE)
+	if !errors.Is(err, unix.EEXIST) {
+		return fmt.Errorf("RENAME_NOREPLACE of %v to %v should have failed with EEXIST, but got %w", f1, f2, err)
+	}
+
+	// RENAME_WHITEOUT will move f3 to f4, leaving a whiteout file behind. This is used internally by overlayfs, but is
+	// not validated before a mount is attempted, as some use cases may not break. Ours will.
+	f3, f4 := filepath.Join(l1, "f3"), filepath.Join(l1, "f4")
+	if err := os.WriteFile(f3, []byte{}, 0o700); err != nil {
+		return err
+	}
+	err = unix.Renameat2(0, f3, 0, f4, unix.RENAME_WHITEOUT)
```

Comment:
The manpage says that `renameat2(RENAME_WHITEOUT)` will fail with EPERM if the caller does not have the CAP_MKNOD capability. Could this result in false negatives (i.e. detect it is unsupported when the kernel and backing fs actually do support it) in Rootless Docker? Perhaps this check should be more limited, considering it a failure iff the error is EINVAL, or considering EPERM to be "inconclusive?" Or is there some sequence of operations we could perform on the overlayfs to indirectly test the filesystem support for RENAME_WHITEOUT without needing the process to have that capability?